File: neet-code/graphs/course-schedule.py
# https://leetcode.com/problems/course-schedule/description/

class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        courseSchedule = []

        inDegree = collections.defaultdict(int)
        graph = collections.defaultdict(list)
        queue = collections.deque()

        for prerequisite in prerequisites:
            prereq, course = prerequisite[0], prerequisite[1]
            graph[prereq].append(course)
            inDegree[course] += 1
        
        for course in range(numCourses):
            if inDegree[course] == 0:
                queue.append(course)
        
        while queue:
            course = queue.popleft()
            courseSchedule.append(course)

            for neighbor in graph[course]:
                inDegree[neighbor] -= 1

                if inDegree[neighbor] == 0:
                    queue.append(neighbor)
        
        return len(courseSchedule) == numCourses
        

File: data-structures-3/topological-sort/courseSchedule.js
function courseSchedule(numCourses, prerequisites) {
  const adjList = Array.from({ length: numCourses }, () => []);
  const inDegree = Array.from({ length: numCourses }, () => 0);

  for (const [course, prereq] of prerequisites) {
    adjList[prereq].push(course);
    inDegree[course] += 1;
  }

  const queue = [];
  for (let node = 0; node < numCourses; node++) {
    if (inDegree[node] === 0) {
      queue.push(node);
    }
  }

  const sortedList = [];
  while (queue.length) {
    const currentNode = queue.shift();
    sortedList.push(currentNode);

    for (const neighbor of adjList[currentNode]) {
      inDegree[neighbor] -= 1;

      if (inDegree[neighbor] === 0) {
        queue.push(neighbor);
      }
    }
  }

  return sortedList.length === numCourses ? sortedList : [];
}
